aws/data_source_aws_instances_test.go

```diff
@@ -113,3 +129,41 @@ data "aws_instances" "test" {
 }
 `, rInt)
 }
+
+func testAccInstancesDataSourceConfig_instance_state_names(rInt int) string {
+	return fmt.Sprintf(`
+data "aws_ami" "ubuntu" {
+  most_recent = true
+
+  filter {
+    name   = "name"
+    values = ["ubuntu/images/hvm-ssd/ubuntu-trusty-14.04-amd64-server-*"]
+  }
+
+  filter {
+    name   = "virtualization-type"
+    values = ["hvm"]
+  }
+
+  owners = ["099720109477"] # Canonical
+}
+
+resource "aws_instance" "test" {
+  count = 2
+  ami = "${data.aws_ami.ubuntu.id}"
+  instance_type = "t2.micro"
+  tags {
+    Name = "TfAccTest-HelloWorld"
+    TestSeed = "%[1]d"
+  }
+}
+
+data "aws_instances" "test" {
+  instance_tags {
+    Name = "*"
```

Comment:
To properly setup the resource and data source ordering, you can reference the `aws_instance` resources created above 😄 

```hcl
Name = "${aws_instance.test.0.tags["Name"]}"
```

This should be enough to delay the data source to catch the two new instances 👍  Let me know if this still is causing an issue and we can dive in deeper. Thanks for your work here!